### PR TITLE
App: reduce QtCore coupling in core App code

### PR DESCRIPTION
**Pull request by tritao**
This PR reduces QtCore/QtXml coupling in core App code by moving several small infrastructure paths to Base and standard-library helpers.

The goal is to allow the core of FreeCAD to be compiled without a dependency on Qt, making it easier to port the core to non-desktop platforms.

### Base infrastructure

- Add byte-view and byte-buffer primitives for App naming code.
- Add SHA1, path utility, and platform-path helpers in Base.
- Keep platform default path discovery separate from FreeCAD-specific directory policy.

### Naming and string hashing

- Move `IndexedName`, `StringHasher`, `MappedName`, and `ElementMap` away from Qt byte/string containers.
- Preserve existing hashing and serialized naming behavior while removing Qt API usage.

### Paths, directories, and temporary files

- Replace App `QDir`/`QFileInfo` usage with `std::filesystem` and Base path helpers.
- Move application directory setup away from Qt path APIs.
- Remove Qt temp-directory usage in safe mode.
- Normalize document path handling through Base path utilities.

### XML and document hashing

- Parse `branding.xml` with Xerces instead of QtXml.
- Drop the App QtXml dependency.
- Replace `QCryptographicHash` usage in document transient directory hashing.

### Translation and program information

- Replace App `tr()` macro usage with `Base::Translation`.
- Make program information support Qt-optional builds.
- Stop relying on generated QtCore headers from core App code.

### Expression tokenizer and build plumbing

- Make `ExpressionTokenizer` Qt-free.
- Add `BUILD_QT` plumbing and gate Qt bridge code in CMake.

### CODE REVIEW

File: cMake/FreeCAD_Helpers/InitializeFreeCADBuildOptions.cmake [hunk 1]
```diff
@@ -5,6 +5,7 @@ macro(InitializeFreeCADBuildOptions)
 
     option(BUILD_FORCE_DIRECTORY "The build directory must be different to the source directory." OFF)
     option(BUILD_GUI "Build FreeCAD Gui. Otherwise you have only the command line and the Python import module." ON)
+    option(BUILD_QT "Build Qt-dependent support code (e.g. App Qt bridges). This is currently required by parts of App even when BUILD_GUI=OFF." ON)
```
Comment (Lgt2x) [inline cMake/FreeCAD_Helpers/InitializeFreeCADBuildOptions.cmake]: This variable (name is confusing btw) could be set automatically depending on whether Qt was found or not, I don't see the case where you'd want to toggle it on or off

File: src/App/Document.cpp [hunk 2]
```diff
@@ -1103,14 +1104,31 @@ std::string Document::getTransientDirectoryName(const std::string& uuid,
 {
     // Create a directory name of the form: {ExeName}_Doc_{UUID}_{HASH}_{PID}
     std::stringstream out;
-    QCryptographicHash hash(QCryptographicHash::Sha1);
-#if QT_VERSION < QT_VERSION_CHECK(6, 3, 0)
-    hash.addData(filename.c_str(), filename.size());
-#else
-    hash.addData(QByteArrayView(filename.c_str(), filename.size()));
-#endif
+    boost::uuids::detail::sha1 sha1;
```
Comment (ElementW) [inline src/App/Document.cpp]: Runs afoul of the goal of removing Boost one day (#26882). I don't think this hash needs to be SHA1 or even a cryptographic hash to begin with, can't we use another algo?

File: src/App/MappedName.h [hunk 3]
```diff
@@ -56,53 +62,64 @@ namespace Data
...
-     * provided, the string must be null-terminated.
-     */
-    explicit MappedName(const char* name, int size = -1);
+    /// Create a MappedName from a C string, optionally prefixed by an element map prefix, which
+    /// will be omitted from the stored MappedName.
+    ///
+    /// \param name The new name. A deep copy is made.
+    /// \param size Optional, the length of the name string. If not provided, the string must be
+    /// null-terminated.
+    explicit MappedName(const char* name, int size = -1)
```
Comment (ElementW) [inline src/App/MappedName.h:89]: Can't these two just forward into a single `std::string_view`-based constructor? The logic is the same, and you still pay the price of `strlen` anyway in the C string case.

File: src/App/Document.cpp (hunk 2 above)
Comment (tritao) [reply src/App/Document.cpp]: I've actually split this PR off from the Boost removal branch, it's next in sequence, so the next PR for Boost removal will fix this.

File: src/App/SafeMode.cpp [hunk 4]
```diff
@@ -23,31 +23,50 @@
...
-        tempDir = nullptr;
+    std::error_code error;
+    const fs::path base = fs::temp_directory_path(error);
+    if (error) {
+        return false;
     }
-    return tempDir;
+
+    std::random_device rd;
+    std::mt19937_64 rng(rd());
```
Comment (ElementW) [inline src/App/SafeMode.cpp]: Spinning up an mt19937 for this is overkill here. Consider doing simpler bit twiddling [like I've done for `Application::uniqueInstanceId()`](https://github.com/FreeCAD/FreeCAD/pull/29962/changes#diff-3bbde2bd74bfaa9e7fbc591ca2dfc76be29f2e4f08358eb4ee7d7c0c2bc3e27bR1344-R1353), especially since your existing code already operates over the full 64 bit range, or using a smaller random engine like `minstd_rand`.

File: cMake/FreeCAD_Helpers/ChooseQtVersion.cmake [hunk 5]
```diff
@@ -47,6 +47,7 @@
 
 macro(ChooseQtVersion)
   set(freecad_supported_qt_versions "Auto" 5 6)
+  set(FREECAD_HAS_QT OFF CACHE INTERNAL "Whether Qt support is available for this build." FORCE)
```
Comment (Lgt2x) [inline cMake/FreeCAD_Helpers/ChooseQtVersion.cmake]: Why `CACHE` + `FORCE`, a simple `set(` would not work here?

File: cMake/FreeCAD_Helpers/SetupQt.cmake [hunk 6]
```diff
@@ -1,5 +1,13 @@
 # -------------------------------- Qt --------------------------------
 
+if(NOT FREECAD_HAS_QT)
+    # FCGlobal.h includes <QtCore.h> unconditionally. This is a generated FreeCAD header (not a Qt header).
+    file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/src)
+    set(HAVE_Q_DISABLE_COPY_MOVE 0)
+    configure_file(${CMAKE_SOURCE_DIR}/src/QtCore.h.cmake ${CMAKE_BINARY_DIR}/src/QtCore.h)
+    return()
+endif()
```
Comment (Lgt2x) [inline cMake/FreeCAD_Helpers/SetupQt.cmake]: Could we make the QtCore include conditional to `FREECAD_HAS_QT`, and inline the content of `QtCore.h.cmake` in FCGlobal.h instead? I'm not a big fan of including a cmake-generated header unconditionally as a system header with `<>`

File: src/App/Document.cpp (hunk 2 above)
Comment (tritao) [reply src/App/Document.cpp]: I ended up just reworking in this PR, now uses SHA1 directly without Boost.

**Comment by tritao** [comment]
Been putting some further cleanups on this, @Lgt2x / @ElementW can you check it again?

File: src/App/SafeMode.cpp [hunk 7]
```diff
@@ -58,10 +77,10 @@ static void _replaceDirs()
...
 
     for (auto const d : dirs) {
-        auto const path = temp_base + "/" + d + "/";
-        auto const qpath = QDir::cleanPath(QString::fromStdString(path));
-        QDir().mkpath(qpath);
-        config[d] = QDir::toNativeSeparators(qpath).toStdString();
+        auto const path = temp_base + PATHSEP + d + PATHSEP;
+        std::error_code error;
+        fs::create_directories(fs::path(temp_base) / d, error);
+        config[d] = path;
```
Comment (ElementW) [inline src/App/SafeMode.cpp]: Why assemble the path twice, once as `fs::path`, once as `string`, instead of only creating it as `fs::path` and converting it to string for `config[d]`?

File: src/App/SafeMode.cpp [hunk 8]
```diff
@@ -58,10 +77,10 @@ static void _replaceDirs()
     };
 
     for (auto const d : dirs) {
-        auto const path = temp_base + "/" + d + "/";
-        auto const qpath = QDir::cleanPath(QString::fromStdString(path));
-        QDir().mkpath(qpath);
-        config[d] = QDir::toNativeSeparators(qpath).toStdString();
+        auto const path = temp_base + PATHSEP + d + PATHSEP;
+        std::error_code error;
```
Comment (ElementW) [inline src/App/SafeMode.cpp]: Error goes unchecked, is temp dir creation failure not fatal? Consider using the throwing variant of `create_directories`.

File: src/App/SafeMode.cpp [hunk 9]
```diff
@@ -74,10 +93,15 @@ void SafeMode::StartSafeMode()
...
+    return !tempDir.empty();
 }
 
 void SafeMode::Destruct()
 {
-    delete tempDir;
+    if (tempDir.empty()) {
+        return;
+    }
+    std::error_code error;
```
Comment (ElementW) [inline src/App/SafeMode.cpp:190]: Temp dir removal failure isn't fatal, but maybe add a comment that you're willingly ignoring `error`'s value.

File: src/App/StringHasher.cpp [hunk 10]
```diff
@@ -51,14 +54,45 @@ using namespace App;
...
+    // boost::hash_combine equivalent (no dependency)
+    return a ^ (b + 0x9e3779b97f4a7c15ULL + (a << 6) + (a >> 2));
+}
+
+Base::ByteBuffer sha1DigestBuffer(Base::BytesView bytes)
+{
+    const auto digest = Base::sha1Digest(bytes);
+    return Base::ByteBuffer::copy(
+        Base::BytesView(reinterpret_cast<const char*>(digest.data()), digest.size()));
+}
```
Comment (ElementW) [inline src/App/StringHasher.cpp]: I'd argue all these could be moved to some public, common utility header instead of remaining private. Especially `hashCombine`.

File: src/Mod/Part/App/TopoShape.h [hunk 11]
```diff
@@ -28,6 +28,8 @@
 #include <list>
 #include <unordered_map>
 
+#include <QVector>
```
Comment (ElementW) [inline src/Mod/Part/App/TopoShape.h:31]: Is there a reason for that one?

File: src/Mod/Part/App/TopoShapeCache.h [hunk 12]
```diff
@@ -35,6 +35,8 @@
 #include <TopTools_IndexedMapOfShape.hxx>
 #include <utility>
 
+#include <QVector>
```
Comment (ElementW) [inline src/Mod/Part/App/TopoShapeCache.h:38]: Same here.

File: src/Mod/Part/App/TopoShape.h (hunk 11 above)
Comment (tritao) [reply src/Mod/Part/App/TopoShape.h:31]: `TopoShape.h` declares methods with `QVector<Data::MappedElement>` parameters, so it must directly include `QVector` instead of relying on a transitive QtCore include.

File: src/Mod/Part/App/TopoShapeCache.h (hunk 12 above)
Comment (tritao) [reply src/Mod/Part/App/TopoShapeCache.h:38]: Same as above.

File: src/App/SafeMode.cpp (hunk 7 above)
Comment (tritao) [reply src/App/SafeMode.cpp]: Fixed.

File: src/App/StringHasher.cpp (hunk 10 above)
Comment (tritao) [reply src/App/StringHasher.cpp]: Done.

File: src/App/SafeMode.cpp (hunk 9 above)
Comment (tritao) [reply src/App/SafeMode.cpp:190]: Done.

File: src/App/SafeMode.cpp (hunk 8 above)
Comment (tritao) [reply src/App/SafeMode.cpp]: Fixed.

File: src/App/StringHasher.cpp [hunk 13]
```diff
@@ -51,14 +54,27 @@ using namespace App;
 
 ///////////////////////////////////////////////////////////
 
+namespace
+{
+
+Base::ByteBuffer sha1DigestBuffer(Base::BytesView bytes)
```
Comment (tritao) [inline src/App/StringHasher.cpp:63]: @ElementW Left this one here for now, but if you think it makes sense to also move it, let me know.

**Comment by tritao** [comment]
@3x380V You had some good feedback on previous related PRs, let me know if you have any thoughts on this one.

**Review by Lgt2x — APPROVED** [review]
Good for me at least on the cmake side. Quickly glanced over the rest of the diff and didn't see anything shocking. Thanks for this effort!

**Comment by 3x380V** [comment]
@tritao, sorry for late reply (and I know I owe replies in some other PRs as well), unfortunately I do not code during summer and reviewing such a nice (at least in by intention and by briefly looking also by implementation) PR is quite time consuming. Something I cannot afford now. Apologies and keep up good work :)
(of course if the PR will still rot there in few months, I'll go thru it carefully)

**Comment by tritao** [comment]
Rebased to fix conflicts.

**Comment by Lgt2x** [comment]
@tritao before I review in depth again, is there something in this PR that could potentially break in the app, and that our QA people could help test and make sure nothing changes?

**Comment by tritao** [comment]
I think the main risk is platform-specific regressions, especially on Windows, due to the changes in the low-level path and temporary-directory handling. 

I've done a further audit on this PR and sent some more commits for hardening things a bit.

File: src/App/ApplicationDirectories.cpp [hunk 14]
```diff
@@ -429,75 +435,39 @@ fs::path ApplicationDirectories::findUserHomePath(const fs::path& userHome)
...
-        dir.mkdir(QStringLiteral("temp"));
-        const QFileInfo fi(dir, QStringLiteral("temp"));
-        userTemp = fi.absoluteFilePath();
+    if (!userHome.empty() && userTemp.empty()) {
+        userTemp = userHome / "temp";
+        if (!Py_IsInitialized()) {
+            try {
+                fs::create_directories(userTemp);
+            } catch (...) {
+            }
```
Comment (Lgt2x) [inline src/App/ApplicationDirectories.cpp]: we may want to do something with the exception here?

File: src/Base/PlatformPaths.cpp [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+    const auto pathEnv = environmentVariableUtf8("PATH");
+    if (!pathEnv) {
+        return {};
+    }
+
+#if defined(FC_OS_WIN32)
+    constexpr char separator = ';';
+#else
+    constexpr char separator = ':';
+#endif
```
Comment (Lgt2x) [inline src/Base/PlatformPaths.cpp]: we have another declaration of separator in `src/Base/interpreter.cpp`, can be worth creating a function returning the platform-dependent separator

File: tests/src/App/PropertyExpressionEngine.cpp [hunk 16]
```diff
@@ -25,6 +26,9 @@ class PropertyExpressionEngineTest: public ::testing::Test
 
     void SetUp() override
     {
+        if (Base::Type::fromName("Sketcher::SketchObject").isBad()) {
+            GTEST_SKIP() << "Sketcher module is unavailable";
+        }
```
Comment (Lgt2x) [inline tests/src/App/PropertyExpressionEngine.cpp]: test could also be refactored to avoid using a sketch object

File: tests/src/App/ProgramInformation.cpp [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+fs::path makeUniqueTempDir()
+{
+    constexpr int maxTries = 128;
+    const fs::path base = fs::temp_directory_path();
+    std::random_device rd;
+    std::mt19937_64 gen(rd());
+    std::uniform_int_distribution<unsigned long long> dist;
+
+    for (int i = 0; i < maxTries; ++i) {
```
Comment (Lgt2x) [inline tests/src/App/ProgramInformation.cpp]: a timestamp-based tempdir creation would eliminate the need for retries

File: src/App/Branding.cpp [hunk 18]
```diff
@@ -22,111 +22,149 @@
...
+    filter.insert("SplashWarningColor");
+
+    filter.insert("StartWorkbench");
+
+    filter.insert("ExeName");
+    filter.insert("ExeVendor");
+    filter.insert("ExeVersion");
+    filter.insert("AppDataSkipVendor");
+    filter.insert("NavigationStyle");
+    filter.insert("UserParameterTemplate");
```
Comment (Lgt2x) [inline src/App/Branding.cpp]: this code is a little silly, the vector can be constant or filled in a loop

File: src/Base/Base64.h [hunk 19]
```diff
@@ -55,7 +55,7 @@ inline std::size_t base64_encode_size(std::size_t len)
 /// Returns the max bytes of a decoded base64 binary string
 inline std::size_t base64_decode_size(std::size_t len)
 {
-    return len / 4 * 3;
+    return len / 4 * 3 + (len % 4 ? 3 : 0);
```
Comment (Lgt2x) [inline src/Base/Base64.h]: I wonder how this was not noticed before

File: src/Base/ByteBuffer.h [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+
+    ByteBuffer() = default;
+    ByteBuffer(const ByteBuffer&) = default;
+    ByteBuffer& operator=(const ByteBuffer&) = default;
+
+    ByteBuffer(ByteBuffer&& other) noexcept
+        : ptr(other.ptr)
+        , len(other.len)
+        , owner(std::move(other.owner))
+    {
```
Comment (Lgt2x) [inline src/Base/ByteBuffer.h]: you can move all of the implementations to a .cpp file

File: src/Base/ByteBuffer.h [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+        if (minCapacity > owner->capacity()) {
+            owner->reserve(minCapacity);
+        }
+        syncFromOwner();
+    }
+
+private:
+    const char* ptr {""};
+    size_type len {0U};
+    std::shared_ptr<std::string> owner;
```
Comment (Lgt2x) [inline src/Base/ByteBuffer.h]: so the actual storage for our ByteBuffer is a string? Why this rather than a STL vector? It is a curious choice

File: src/Base/BytesView.h [hunk 22]
```diff
@@ -1,45 +1,31 @@
...
-TranslateManip::TranslateManip()
+namespace Base
 {
-    SO_NODE_CONSTRUCTOR(TranslateManip);
 
-    auto myDrag = new SoTranslate2Dragger;
-    setDragger(myDrag);
-}
+/// Non-owning view over a byte sequence (may contain embedded NULs).
+using BytesView = std::string_view;
```
Comment (Lgt2x) [inline src/Base/BytesView.h]: same here, I'm not convinced that using string storage for bytes is the best we can do. `std::span` is made for non-owned memory, not necessarily text, and will not expose unknowingly string APIs.

File: src/Base/HashUtils.h [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <cstddef>
+#include <cstdint>
+
+#include <Base/BytesView.h>
+
+namespace Base
+{
+
+inline std::size_t fnv1a64(BytesView bytes) noexcept
```
Comment (Lgt2x) [inline src/Base/HashUtils.h:34]: doc the function, so people know it's a hash function that can be used for many things

File: src/Base/PathUtils.h [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+#endif
+
+namespace Base
+{
+
+struct BaseExport NormalizePathOptions
+{
+    bool makeAbsolute {true};
+    bool weaklyCanonical {true};
+    bool createParentDirectories {false};
```
Comment (Lgt2x) [inline src/Base/PathUtils.h]: I don't see what creating parent dirs have to do with path manipulation utilities, this should be a separate function

File: src/Base/PathUtils.h [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+# include <FCGlobal.h>
+#endif
+
+namespace Base
+{
+
+struct BaseExport NormalizePathOptions
+{
+    bool makeAbsolute {true};
+    bool weaklyCanonical {true};
```
Comment (Lgt2x) [inline src/Base/PathUtils.h]: one option has a verb (action), one does not. can be made `absolute` and `weaklyCanonical`

File: src/App/Document.cpp [hunk 26]
```diff
@@ -1949,32 +1979,17 @@ bool Document::saveToFile(const char* filename) const
...
-            fs::create_directories(parentPath);
+        Base::NormalizePathOptions options;
+        options.makeAbsolute = true;
+        options.weaklyCanonical = true;
+        options.createParentDirectories = true;
 
-            return std::string(filename);
+        const auto normalized =
+            Base::normalizePath(Base::pathFromUtf8(filename ? std::string_view(filename) : std::string_view()), options);
+        if (!normalized) {
```
Comment (Lgt2x) [inline src/App/Document.cpp]:
```suggestion
        if (!normalized.has_value()) {
```

File: src/App/Document.cpp [hunk 27]
```diff
@@ -1078,14 +1104,19 @@ std::string Document::getTransientDirectoryName(const std::string& uuid,
...
+    auto hexChar = [](unsigned char nibble) -> char {
+        return nibble < 10 ? static_cast<char>('0' + nibble) : static_cast<char>('a' + (nibble - 10));
+    };
+
+    std::array<char, 6> hashPrefix {};
+    for (int i = 0; i < 3; ++i) {
+        hashPrefix[i * 2 + 0] = hexChar(static_cast<unsigned char>((digestBytes[i] >> 4U) & 0x0F));
+        hashPrefix[i * 2 + 1] = hexChar(static_cast<unsigned char>((digestBytes[i] >> 0U) & 0x0F));
+    }
     out << Application::getUserCachePath() << Application::getExecutableName() << "_Doc_"
```
Comment (Lgt2x) [inline src/App/Document.cpp]: move this logic to HashUtils or Sha1 and test it, it should not be here

**Review by Lgt2x — COMMENTED** [review]
I'm halfway through the review, with a batch of comments. More general question
1. how do we ensure that the Qt-less build stays in a non-broken state?
2. What's the easiest way to try it, using a container environment?
3. Should CI have a job for it?

File: src/Base/Base64.h (hunk 19 above)
Comment (ElementW) [reply src/Base/Base64.h]:
```suggestion
    return (len + 3) / 4 * 3;
```
Branchless version.

Comment (tritao) [reply src/Base/Base64.h]: The branchless suggestion is actually unsafe, when len is `std::numeric_limits<std::size_t>::max()`.

Seems unlikely to actually happen... but better go with the safer approach.

I've replaced with a more readable alternative based on @Lgt2x suggestion.

Comment (tritao) [reply src/Base/Base64.h]: There were also some other potential overflows in this area, I've hardened it and added couple more edge-case tests.

**Comment by tritao** [comment]
Hey @Lgt2x, thanks for the in-depth review and for catching some of the questionable things in this.

I've updated the PR, should now address all the raised issues.

> I'm halfway through the review, with a batch of comments. More general question
> 
> 1. how do we ensure that the Qt-less build stays in a non-broken state?
> 3. Should CI have a job for it?

My idea was to have an Emscripten CI build target, which would cover this functionality.

But for the time being, I think we should add a Qt-less CI build and maybe disable/remove it once we have Emscripten CI coverage, just to ease the pressure on CI if that's a concern.

> 2. What's the easiest way to try it, using a container environment?

```bash
docker run --rm -it -v "$PWD":/src -w /src ubuntu:24.04 bash
```

Inside the container:

```bash
apt-get update
apt-get install -y sudo
./package/ubuntu/install-apt-packages.sh

cmake -S . -B build/core-no-qt \
  -D CMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE \
  -G Ninja --preset core-no-qt

cmake --build build/core-no-qt -j"$(nproc)" \
  --target FreeCADBase FreeCADApp Base_tests_run App_tests_run

build/core-no-qt/tests/Base_tests_run
build/core-no-qt/tests/App_tests_run
```

Something like that should do it.

File: src/Base/Base64.h (hunk 19 above)
Comment (ElementW) [reply src/Base/Base64.h]: The overflow concern is correct but misplaced. It is *impossible* to have a chunk of memory that's max size_t long on any platform with a userspace, and given the purpose of this function any number being passed in that's higher than `size_t_max*4/7` is invalid. If you get passed a value that large at this point in the code, the overflow is a non-issue because the allocations both before and after are guaranteed to fail.

File: src/Base/ByteBuffer.cpp [hunk 28]
```diff
@@ -0,0 +1,284 @@
...
+ByteBuffer& ByteBuffer::operator=(ByteBuffer&& other)
+{
+    if (this == &other) {
+        return *this;
+    }
+
+    // A borrowed view can point into this buffer's current owner. Make a
+    // stable copy before releasing that owner during the assignment.
+    if (!other.owner && owner && rangesOverlap(ptr, len, other.ptr, other.len)) {
+        owner = std::make_shared<std::vector<std::byte>>();
```
Comment (3x380V) [inline src/Base/ByteBuffer.cpp:64]: owner replaced, freeing the vector, before copying from other.ptr, which points into it.
```
        auto copied = std::make_shared<std::vector<std::byte>>(other.ptr, other.ptr + other.len);
        owner = std::move(copied);
```

File: src/App/StringHasher.cpp [hunk 29]
```diff
@@ -51,14 +57,61 @@ using namespace App;
...
+        }
+    }
+
+    std::vector<std::byte> result;
+    const std::size_t consumed = Base::base64_decode(result, encoded.data(), encoded.size());
+    if (consumed != dataLength) {
+        return false;
+    }
+
+    decoded = Base::ByteBuffer::copy(Base::asBytes(result.data(), consumed));
```
Comment (3x380V) [inline src/App/StringHasher.cpp:100]: consumed (input chars) used as the output length yields to trailing NULs.
```decoded = Base::ByteBuffer::copy(Base::asBytes(result.data(), result.size()));```
(Discovered by running `make test`)

File: src/App/StringHasher.h [hunk 30]
```diff
@@ -309,314 +383,287 @@ class AppExport StringID: public Base::BaseClass, public Base::Handled
...
-        if (_sid) {
-            return _sid->value();
-        }
-        return 0;
+inline const char* StringIDRef::constData() const
+{
+    if (_sid) {
+        assert(_index == 0);
+        assert(_sid->postfix().empty());
+        return reinterpret_cast<const char*>(_sid->data().data());
```
Comment (3x380V) [inline src/App/StringHasher.h:573]: I removed this altogether, nowhere used except the test, which spotted it returns a const char* into non-terminated bytes.

**Comment by 3x380V** [comment]
I'm sorry to be late to the party... Tried to run test, didn't work:
- tests/src/App/MappedName.cpp - does not build; now both sides cast to `const void*` for the pointer-identity checks.
- tests/src/App/ComplexGeoData.cpp - both `elementType` tests fixed.
- tests/src/App/StringHasher.cpp — added `toStdString` and `rawData` helpers, then content checks go through `toQByteArray(...) == qba` and identity checks through `const void*` comparison, replacing `EXPECT_STREQ`/`EXPECT_STRNE` on `constData()`.

The rest of remarks above, branch rebased with conflicts and the above problems fixed here: https://github.com/3x380V/FreeCAD/tree/app-qt-free
(also fixed one spot adding wrong line endings into `Document.cpp`, but `main` still contains more as noted in https://github.com/FreeCAD/FreeCAD/issues/29603#issuecomment-5573249578_)

**Comment by 3x380V** [comment]
Using this branch for daily work I start occasionally seeing
```
Reading failed from embedded file: StringHasher.Table.txt (5120 bytes, 1451 bytes compressed)
<ElementMap> <string>(1): Missing element name prefix id
```
Edit: above is seen also in the `main` branch...

**Comment by maxwxyz** [comment]
@tritao PR has some conflicts

**Comment by 3x380V** [comment]
@tritao, conflicts and issues I pointed out still present...

**Comment by tritao** [comment]
@3x380V Thanks for the review, much appreciated.

I have now rebased this to latest main, and think all your issues are addressed.

Please let me know if I missed anything.

**Comment by 3x380V** [comment]
@tritao, good, now both our branches are rebased onto 83df87da14. And I guess your branch won't compile. You might want compare against https://github.com/3x380V/FreeCAD/tree/app-qt-free
Btw, I'm using your work on daily basis, no problems found so far (well, spotted few, but none related to your commits)

**Comment by 3x380V** [comment]
@tritao, branch has conflicts again and #32749 is to blame. Let me therefore resolve conflicts again: https://github.com/3x380V/FreeCAD/tree/app-qt-free
For testing purposes I also locally rebased this PR on top of #32716 which I created solely to make tracking regression here easier. Two new warnings pop out, both signed vs unsigned compare, fixups are provided: 973e76bcacf34e5a356e2b3f470b6c4ef15a4fc8 and 4ea298ed60535a9d1c8bec7d6480120b67933a9d

**Comment by tritao** [comment]
Thanks @3x380V, rebased and fixed.

**Comment by 3x380V** [comment]
It seems you did rebase on your own and that reintroduced else-after-return blocks in [App: reduce QtCore coupling in core App code](https://github.com/FreeCAD/FreeCAD/pull/30839/changes/2e0d515500c2c1955343a6bce5c5c241b26f021b) which were fixed by #32749

**Comment by 3x380V** [comment]
I force pushed my branch again, to make comparing easier. There still is a difference which autoformat should correct (it is not yet enabled on App)
```diff
--- a/src/App/PropertyLinks.cpp
+++ b/src/App/PropertyLinks.cpp
@@ -5606,11 +5606,9 @@ bool PropertyXLinkSubList::upgrade(Base::XMLReader& reader, const char* typeName
         setValues(linkProp.getValues());
         return true;
     }
-    if (
-        typeName == PropertyLinkSubListGlobal::getClassTypeId().getName()
+    if (typeName == PropertyLinkSubListGlobal::getClassTypeId().getName()
         || typeName == PropertyLinkSubList::getClassTypeId().getName()
-        || typeName == PropertyLinkSubListChild::getClassTypeId().getName()
-    ) {
+        || typeName == PropertyLinkSubListChild::getClassTypeId().getName()) {
         PropertyLinkSubList linkProp;
         linkProp.setContainer(getContainer());
         linkProp.Restore(reader);
```

**Comment by 3x380V** [comment]
@tritao, I rebased my branch with conflicts resolution, in case you want to save some time doing it on your own. I was also using this branch for past two weeks without any issues for my daily work without noticing any regression, thou I need only PartDesign (with Assembly4) and TechDraw.

**Comment by tritao** [comment]
Rebased.

**Comment by 3x380V** [comment]
@tritao, I restored `ChooseQtVersion.cmake` after Qt5 support removal, so it still selects between none and Qt6 and makes FC ready for eventual Qt7. Also conflicts after std::format merge were resolved and you can find all that in the usual place. I have additional work queued on top of this PR, so looking forward to see it merged.

**Comment by 3x380V** [comment]
.oO(Diffing against https://github.com/3x380V/FreeCAD/tree/app-qt-free before force-pushing would still save some time here...)

File: src/App/Document.cpp [hunk 31]
```diff
@@ -1245,11 +1241,10 @@ void Document::Restore(Base::XMLReader& reader)
 void DocumentP::checkStringHasher(const Base::XMLReader& reader)
 {
     if (reader.hasReadFailed("StringHasher.Table.txt")) {
-        Base::Console().error(QT_TRANSLATE_NOOP(
-            "Notifications",
+        Base::Console().error(
             "\nIt is recommended that the user right-click the root of "
```
Comment (3x380V) [inline src/App/Document.cpp:1245]: This drops _context_ and also `lupdate` will not find the string. Just hit this while continuing de-Qt on Sketcher/App and Part/App.

**Comment by 3x380V** [comment]
There still is `fmt::format` in src/App/ApplicationDirectories.cpp, so it will not link.

File: src/FCGlobal.h [hunk 32]
```diff
@@ -83,3 +83,7 @@
 #define FC_DEFAULT_COPY_MOVE(Class) \
     FC_DEFAULT_COPY(Class) \
     FC_DEFAULT_MOVE(Class)
+
+#ifndef HAVE_Q_DISABLE_COPY_MOVE
+#define Q_DISABLE_COPY_MOVE FC_DISABLE_COPY_MOVE
+#endif
```
Comment (3x380V) [inline src/FCGlobal.h]: There is now nothing to define `HAVE_Q_DISABLE_COPY_MOVE`:
```
In file included from /home/ladis/src/FreeCAD/src/Gui/Action.h:31,
                 from /home/ladis/src/FreeCAD/src/Gui/CommandPyImp.cpp:29:
/home/ladis/src/FreeCAD/src/FCGlobal.h:88:9: warning: "Q_DISABLE_COPY_MOVE" redefined
   88 | #define Q_DISABLE_COPY_MOVE FC_DISABLE_COPY_MOVE
      |         ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/qt6/QtCore/qtdeprecationmarkers.h:8,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qtcoreexports.h:9,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qglobal.h:26,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qregularexpression.h:9,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/QRegularExpression:1,
                 from /home/ladis/src/FreeCAD/src/Gui/CommandPyImp.cpp:24:
/usr/include/x86_64-linux-gnu/qt6/QtCore/qtclasshelpermacros.h:27:9: note: this is the location of the previous definition
   27 | #define Q_DISABLE_COPY_MOVE(Class) \
      |         ^~~~~~~~~~~~~~~~~~~

```
